presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -368,23 +370,46 @@ private JoinEnumerationResult setJoinNodeProperties(JoinNode joinNode)
             if (isAtMostScalar(joinNode.getLeft(), lookup)) {
                 return createJoinEnumerationResult(joinNode.flipChildren().withDistributionType(REPLICATED));
             }
-
-            List<JoinEnumerationResult> possibleJoinNodes = new ArrayList<>();
-            JoinDistributionType joinDistributionType = getJoinDistributionType(session);
-            if (joinDistributionType.canPartition() && !joinNode.isCrossJoin()) {
-                possibleJoinNodes.add(createJoinEnumerationResult(joinNode.withDistributionType(PARTITIONED)));
-                possibleJoinNodes.add(createJoinEnumerationResult(joinNode.flipChildren().withDistributionType(PARTITIONED)));
-            }
-            if (canReplicate(joinNode, context)) {
-                possibleJoinNodes.add(createJoinEnumerationResult(joinNode.withDistributionType(REPLICATED)));
-                possibleJoinNodes.add(createJoinEnumerationResult(joinNode.flipChildren().withDistributionType(REPLICATED)));
-            }
+            List<JoinEnumerationResult> possibleJoinNodes = getPossibleJoinNodes(joinNode, getJoinDistributionType(session));
+            verify(!possibleJoinNodes.isEmpty(), "possibleJoinNodes is empty");
             if (possibleJoinNodes.stream().anyMatch(UNKNOWN_COST_RESULT::equals)) {
                 return UNKNOWN_COST_RESULT;
             }
             return resultComparator.min(possibleJoinNodes);
         }
 
+        private List<JoinEnumerationResult> getPossibleJoinNodes(JoinNode joinNode, JoinDistributionType distributionType)
+        {
+            checkArgument(joinNode.getType() == INNER, "unexpected join node type: %s", joinNode.getType());
+
+            if (joinNode.isCrossJoin()) {
+                return getPossibleJoinNodes(joinNode, REPLICATED);
+            }
+
+            switch (distributionType) {
+                case PARTITIONED:
```

Comment:
Could you add a test for that?

Reply:
There is a test for that actually, `testRepartitionsWhenRequiredBySession`. It used to work before. It didn't work only in case the replicated limit was set, and the replicated join type was explicitly set.